core/src/main/scala/com/salesforce/op/utils/stages/NameDetectUtils.scala

```diff
@@ -0,0 +1,377 @@
+) extends JsonLike
+
+private[op] case class GenderStats(numMale: Int = 0, numFemale: Int = 0, numOther: Int = 0) extends JsonLike
+
+/**
+ * Defines the case class monoid that will accumulate stats on name detection in a single pass over the data
+ * @param guardCheckQuantities     a GuardCheckStats object that uses Algebird approximate algorithms to compute stats
+ * @param dictCheckResult          an Algebird AveragedValue object to automatically compute the percentage of name
+ *                                 tokens per entry, averaged over all rows
+ * @param genderResultsByStrategy  a map from the serialized GenderDetectStrategy to GenderStats case class;
+ *                                 the `numOther` value will be used to sort and find the best gender detection strategy
+ */
+private[op] case class NameDetectStats
+(
+  guardCheckQuantities: GuardCheckStats,
+  dictCheckResult: AveragedValue,
+  genderResultsByStrategy: Map[String, GenderStats]
+) extends JsonLike {
+  // TransmogrifAI's JsonUtils doesn't play nice with HLL
+  override def toJson(pretty: Boolean): String = {
+    val result: Map[String, Any] = Map(
+      "guardCheckQuantities" -> Map(
+        "countBelowMaxNumTokens" -> this.guardCheckQuantities.countBelowMaxNumTokens,
+        "countAboveMinCharLength" -> this.guardCheckQuantities.countAboveMinCharLength,
+        "approxMomentsOfTextLength" -> this.guardCheckQuantities.approxMomentsOfTextLength,
+        "approxNumUnique" -> {
+          val hllMonoid = new HyperLogLogMonoid(NameDetectUtils.HLLBits)
+          hllMonoid.sizeOf(this.guardCheckQuantities.approxNumUnique).toString
+        }
+      ),
+      "dictCheckResults" -> this.dictCheckResult,
+      "genderResultsByStrategy" -> this.genderResultsByStrategy
+    )
+    JsonUtils.toJsonString(result, pretty = pretty)
+  }
+}
+private[op] case object NameDetectStats {
+  def monoid: Monoid[NameDetectStats] = new Monoid[NameDetectStats] {
+    // Ideally, we could have avoided defining all of this
+    // but Algebird's case class macro is not documented (at all) and spotty
+    override def plus(l: NameDetectStats, r: NameDetectStats): NameDetectStats = NameDetectStats(
+      GuardCheckStats(
+        l.guardCheckQuantities.countBelowMaxNumTokens + r.guardCheckQuantities.countBelowMaxNumTokens,
+        l.guardCheckQuantities.countAboveMinCharLength + r.guardCheckQuantities.countAboveMinCharLength,
+        MomentsGroup.plus(
+          l.guardCheckQuantities.approxMomentsOfTextLength,
+          r.guardCheckQuantities.approxMomentsOfTextLength
+        ),
+        l.guardCheckQuantities.approxNumUnique + r.guardCheckQuantities.approxNumUnique
+      ),
+      l.dictCheckResult + r.dictCheckResult,
+      l.genderResultsByStrategy + r.genderResultsByStrategy
+    )
+    override def zero: NameDetectStats = NameDetectStats.empty
+  }
+
+  def empty: NameDetectStats = NameDetectStats(GuardCheckStats(), AveragedGroup.zero, Map.empty[String, GenderStats])
```

Comment:
I thought this would be easy to do but then quickly realized that the whole reason I have to implement a custom `toJson` is because Algebird doesn't serialize nicely yet (https://github.com/twitter/algebird/issues/119). So, I don't have an easy way of deserializing HLL objects, and I also won't be using a `fromJson` method for this PR (the `toJSON` is just for printing/debugging, and I've made that more clear in the comment now).

Reply:
I think there is a way to serialize algebird types with Kryo. It should be somewhere in the code already.